mixins/provider-mixin.js

```diff
@@ -0,0 +1,38 @@
+export const ProviderMixin = superclass => class extends superclass {
+
+	constructor() {
+		super();
+		this._providers = new Map();
+		this.addEventListener('d2l-request-provider', this._handleRequest);
+	}
+
+	_handleRequest(e) {
+		if (this._providers.has(e.detail.key)) {
+			e.detail.provider = this._providers.get(e.detail.key);
+			e.stopPropagation();
+		}
+	}
+
+	provide(key, factory) {
+		this._providers.set(key, factory);
+	}
+
+	delete(key) {
+		this._providers.delete(key);
+	}
+};
+
+export const RequestProviderMixin = superclass => class extends superclass {
+
+	requestProvider(key) {
+		const event = new CustomEvent('d2l-request-provider', {
+			detail: { key },
+			bubbles: true,
+			composed: true,
```

Comment:
Weird that Justin's example wasn't `composed`... but it won't cross shadow DOM boundaries otherwise so... yeah.

Reply:
I can honestly say I didn't even notice that, but that's true... maybe it was a "presentation doesn't match actual code" scenario?